src/strings/ops.c

```diff
@@ -1630,16 +1631,35 @@ MVMString * MVM_string_join(MVMThreadContext *tc, MVMString *separator, MVMObjec
     is_str_array = REPR(input)->pos_funcs.get_elem_storage_spec(tc,
         STABLE(input)).boxed_primitive == MVM_STORAGE_SPEC_BP_STR;
 
+    /* If there's only one element to join, just return it. */
+    if (elems == 1) {
+        if (is_str_array) {
+            MVMString *piece = MVM_repr_at_pos_s(tc, input, 0);
+            if (piece)
+                return piece;
+        }
+        else {
+            MVMObject *item = MVM_repr_at_pos_o(tc, input, 0);
+            if (item && IS_CONCRETE(item))
+                return MVM_repr_get_str(tc, item);
+        }
+    }
+
     /* Allocate result. */
     MVMROOT(tc, separator, {
     MVMROOT(tc, input, {
         result = (MVMString *)MVM_repr_alloc_init(tc, tc->instance->VMString);
     });
     });
 
+    /* If the separator is a strand, find out its storage_type. */
+    all_strands_of_same_type = separator->body.storage_type == MVM_STRING_STRAND;
+    if (all_strands_of_same_type)
+        strand_type = separator->body.storage.strands->blob_string->body.storage_type;
```

Comment:
Why does it matter that all strands have the same storage type? A strand string can reference one 8-bit and another 32-bit wide strand just fine.

Reply:
Huh, I think I misinterpreted something @timo said (https://irclog.perlgeek.de/perl6-dev/2017-09-23#i_15206788) about doing that as an optimization. I thought he meant make support for mixed strings, but now I see he meant actually make strings with mixed strands, since that is actually supported. Cool, I'll add a commit removing that restriction.